### ISSUE TITLE
August Battery Level

### ISSUE THREAD

**Issue by hongkongkiwi** [mtvg/August#5]
I would like to read the current battery level in status and cached. Do you have any idea how I could get this measurement? (does your current code support it? I couldn't see it).

**Comment by mtvg** [comment]
The code doesn't support that, and I don't think the device itself provides that level of information. The only information available seems to be whether or not the battery is low.

Here is a capture of opening the August app and unlocking the door. You can see that the battery level is returned after unlocking. But its value is 0 when my battery is not low. 

	71	SWRITE	0100000085a818027abb75e1009c711c0f01	-> SEC_LOCK_TO_MOBILE_KEY_EXCHANGE
	74	SREAD	0200000074f58018a5386daee5d111390f01	<- SEC_LOCK_TO_MOBILE_KEY_EXCHANGE
	79	SWRITE	03000000bdeb4ac351ad2be3ef6689590f01	-> SEC_INITIALIZATION_COMMAND
	83	WRITE	ee02000c0200000000000000000000000200	-> LOCK_STATUS (0x2 LOCK_STATE)
	88	READ	bb02003a0200000005000000000000000200	<- LOCK_STATUS (0x2 LOCK_STATE) = 0x5
	91	WRITE	ee0200e52900000000000000000000000200	-> LOCK_STATUS (0x29 GIT_HASH)
	99	WRITE	ee0200e22c00000000000000000000000200	-> LOCK_STATUS (0x2c undefined)
	106	READ*	bd24982e4b9d5cb8125824d55763ed870200	
	109	WRITE	ee0200050900000000000000000000000200	-> LOCK_STATUS (0x9 LOCK_EVENTS_UNREAD)
	361	WRITE	ee0a00060000000000000000000000000200	-> UNLOCK
	368	READ*	f0baf419944711147a30fa56a52e37360200	
	591	READ	bb0a00390000000000000000000000000200	<- UNLOCK
	594	READ	bb02008d0223000003000000a0170020a017	<- LOCK_STATUS (0x2302 undefined) = 0x3
	597	WRITE	ee02000c0200000000000000000000000200	-> LOCK_STATUS (0x2 LOCK_STATE)
	601	WRITE	ee0200040a00000000000000000000000200	-> LOCK_STATUS (0xa RTC)
	608	READ	bb02003c0200000003000000000000000200	<- LOCK_STATUS (0x2 LOCK_STATE) = 0x3
	610	READ	bb0200180a000000b027f256000000000200	<- LOCK_STATUS (0xa RTC) = 0x56f227b0
	613	WRITE	ee1000e2af27f25600000000000000000200	
	617	WRITE	ee0200050900000000000000000000000200	-> LOCK_STATUS (0x9 LOCK_EVENTS_UNREAD)
	618	READ	aa100026af27f25600000000000000000200	
	622	READ	bb0200370900000001000000000000000200	<- LOCK_STATUS (0x9 LOCK_EVENTS_UNREAD) = 0x1
	625	WRITE	ee0000100000000000000000000000000200	-> GET_LOCK_EVENTS
	634	READ	bb00007fb027f256000301010000a0000200	<- GET_LOCK_EVENTS{"timestamp":1458710448,"opcode":0,"currentLockState":3,"rssi":"01010000a0","error":0}
	636	READ	bb00014ae414000000000000000000000200	<- GET_LOCK_EVENTS{"currentAngularPosition":5348,"targetAngularPosition":0,"coulombCounter":0,"currentSamples":[0,0]}
	638	READ	bb0002410000000000000000000000000200	<- GET_LOCK_EVENTS{"currentSamples":[0,0,0,0,0,0]}
	640	READ	bb00037b0000000000002214870800000200	<- GET_LOCK_EVENTS{"currentSamples":[0,0],"batteryLevel":0}

I'll try to see what's the returned value when my battery is low, and potentially add that feature!

**Comment by hongkongkiwi** [comment]
Thanks!

That would be really useful. I'm taking your library and integrating it
with smart things so knowing if the battery is low would be awesome

Pity they don't have the battery level when not low.

How about the door lock event history? Do you know if it is stored on the
lock?

>

-- 
Sent from Gmail Mobile

**Comment by mtvg** [comment]
The event history is stored on August web servers. If you want to track any action that is performed by the official app, you would have to access their private API... But I haven't find a way to get access to that yet, they seem to use an SSL pinning mechanism, I'll see if I can get around that.

**Comment by hongkongkiwi** [comment]
Ah I see, it's a shame but makes sense.

I think if you just replace the public cert they are using with your own
one and recompile the Android app that would get around pinning and you
could see their API calls. Let me have a look at this too.

Pinning is only useful with an in app referenced public cert to pin against.

P.S. What are you using for Bluetooth sniffing?

On Friday, 25 March 2016, Mathieu Vignau <notifications@github.com> wrote:

> The event history is stored on August web servers. If you want to track
> any action that is performed by the official app, you would have to access
> their private API... But I haven't find a way to get access to that yet,
> they seem to use an SSL pinning mechanism, I'll see if I can get around
> that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mtvg/August/issues/5#issuecomment-201175539>
>


-- 
Sent from Gmail Mobile

**Comment by mtvg** [comment]
Cool, I'm on iPhone though, but I want to try using this on a jailbroken device: https://github.com/iSECPartners/ios-ssl-kill-switch

For the BLE sniffing I'm using this little handy device: https://www.adafruit.com/products/2269 Then, I'm using this software https://sourceforge.net/projects/nrfblesnifferosx/ on my mac to sniff to Wireshark. 
If you are on Android, you can sniff directly from the device using the developer option to [enable Bluetooth HCI snoop](https://www.nowsecure.com/blog/2014/02/07/bluetooth-packet-capture-on-android-4-4/)

Then I can decrypt August's secure protocol from the pcap file using this: https://github.com/sretlawd/augustctl/blob/master/tools/decode_capture.js

**Comment by tySwift93** [comment]
@hongkongkiwi @mtvg 
I actually figured out how to read the August app's API calls awhile ago and have been using it in my own SmartThings device handler, which makes calls pretty much directly to the August Connect.  Works flawlessly.  

I just tested using this: https://github.com/nabla-c0d3/ssl-kill-switch2 and by enabling it and then using Charles Proxy (or anything else that lets you sniff traffic travelling to and from your iPhone) I was able to see the exact same calls coming from the August app that I had manually discovered a few weeks ago using Flex Loader and sifting through thousands of lines of data...

Long story short, it's possible!

**Comment by tySwift93** [comment]
By the way, I looked into what kind of info the server has about battery.  It does have an actual value in the responses, however it does not have anything to do with the actual lock's battery.  The server seems to just guess at how much battery is left based on when you first installed fresh batteries, which resets the value to 100.00000000000.  Every time that I make a new call, the number decreases by like 0.0000000001.  This is even after I pulled the batteries from the lock and kept checking the server's responses.

Basically, the app/server also don't know how much battery is actually left, it just makes a best guess.

**Comment by hongkongkiwi** [comment]
@tySwift93 are you able to share your connect code? That would be pretty handy!!

So strange about the battery, can't really understand why they did it that way. That totally explains why sometimes the app keeps telling me my battery is low even though I know it's fine.

@mtvg Thanks so much! I'll get a Bluetooth sniffer so I can help :)

**Comment by tySwift93** [comment]
Sure I can!  I am not very good at writing device types, so everything is hard coded.  Also, in order to protect my lock info (all the little bits that make it send commands specifically to MY lock), I will replace those values with X's and write a small comment that explains how I figured the value out for my own system.  What's the best way of sharing the code with you?

**Comment by hongkongkiwi** [comment]
@tySwift93 you can email me a dropbox link if you want andy@savage.hk . I'm still learning groovy, but lets see if I can make a generic version. 

Using mtvg's code I'm also working on a server to run on my Raspberry Pi to achieve the same result.

**Comment by tySwift93** [comment]
Sent! Let me know what you think.
And I'd be interested to know if you use it for anything cool!

**Comment by mtvg** [comment]
Nice! It makes a lot of sense to use directly August Connect, would love to see that code too if you don't mind: mat@mtvg.net

I'm still gonna work on this, I'm controlling multiple BLE devices with my Pi, and like the option to use only one device to do so. 

@hongkongkiwi I own an August Connect though, and one of the missing feature of this project is to be able to get notified when someone manually opened the door, I'll try to sniff that and integrate it in this project.

**Comment by tySwift93** [comment]
@mtvg Will send the code to you right now!  It will work perfectly if you add in the missing bits of data (explained in the code comments) and change the section that makes the device type automatically "re-lock" itself.  You'll see what I mean when you look at the code.

If you figure out how to detect manual unlocks that would be incredible!  Also, I'm assuming that if you could get manual unlocks, then you could probably also get unlocks that happened because a code was entered on the August Keypad?  Would love to know if you find anything out about manual/code unlocks.

Just a thought, could you sniff the connection between the Connect and the lock when the Connect sends an unlock/lock/status command to the lock?  If so, could a server be created that maintains a constant connection to the Connect, which in turn does the sending of commands to the lock?  Hopefully that makes sense.

**Comment by hongkongkiwi** [comment]
@tySwift93 - This wouldn't really work because what if somebody unlocks using Bluetooth without going through connect, then you wouldn't necessarily know if it's been unlocked or not.

Using the app API, the best way is to poll the lock events list, this way your not draining the battery and at the same time anytime it is unlocked using connect or the app it should show up in the lock events list. If you poll that every minute, you should know (within a minute whether it's unlocked or not). 

Better to interact with August's servers rather than try and intercept locally. After all your essentially just emulating what an app does.

**Comment by mxkr** [comment]
@tySwift93 great job on figuring out the connect API! If you don't mind I would love to take a look as well: mxkr00@gmail.com
Thanks!

@mtvg did to you manage to sniff on the Connect? I am looking for the message the connect sends to the lock to dump the recent lock events. Since BLE devices only allow single connection, the connect is not actually connected to the lock all the time, and the lock must be broadcasting to the connect manual unlocks or more likely notifying it to request the recent lock events.
Snooping on my phone I found this call: EE2D00E3000000000000000000000000
It returns epoch time, lock status (3 or 5) and I think the offline key offset, but I couldn't make too consistent of an expected return value.

@hongkongkiwi do you see the manual lock/unlock events with the server events list? I know for sure they log it. A few months ago they accidentally let them display on the activity feed, but they told me that's a bug and they don't display it because the timer in the lock is not reliable enough.
I also started getting unlock notifications for keypad use by guests, actually mentioning their name as if they used they phone. Nothing in the activity log though. Supposedly only the Homekit enabled lock supports this, so it must be another leak/bug.

**Comment by tySwift93** [comment]
@mxkr Sent!
